### PR TITLE
CLI wallet: avoid directly overwriting wallet file on exit #1109

### PR DESCRIPTION
**Pull request by cogutvalera**
PR for #1109

### CODE REVIEW

**Comment by abitmore** [comment]
The logic looks fine, although naming new file as ".backup" is not so good.

**Comment by cogutvalera** [comment]
we should create another issue for fc::ofstream IMHO

**Comment by abitmore** [comment]
> AFAICS fc::ofstream does not throw nor otherwise communicate failure on write(), flush() or close().

It means we need to make sure the new (tmp) file is good before calling `fc::rename`, otherwise it's still possible to lose data. @cogutvalera creating a new issue is fine, but I think this issue/PR depends on the fix of it.

**Comment by cogutvalera** [comment]
review lost commits, now checking `tmp` wallet file contents to be equal as required data before renaming this `tmp` wallet file

File: libraries/wallet/wallet.cpp [hunk 1]
```diff
@@ -808,10 +808,25 @@ class wallet_api_impl
...
+
+         std::string wallet_file_content;
+         fc::read_file_contents(wallet_filename, wallet_file_content);
+
+         if (wallet_file_content == data) {
+            fc::rename( wallet_filename + ".tmp", wallet_filename );
+         } 
+         else 
+         {
+            disable_umask_protection();
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: This is not necessary, because it already happens in the outer try/catch.

File: libraries/wallet/wallet.cpp [hunk 2]
```diff
@@ -808,10 +808,25 @@ class wallet_api_impl
...
+            throw;
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: Please throw an fc::exception subtype with an appropriate error message.

File: libraries/wallet/wallet.cpp (hunk 1 above)
Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: Why it is not necessary ? I've tested & checked it and we won't get into outer try/catch code after throwing exception here that's why I've placed `disable_umask_protection`

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: Then we also need to throw `fc::exception` subtype in outer try/catch block ? right ?

File: libraries/wallet/wallet.cpp (hunk 1 above)
Comment (pmconrad) [reply libraries/wallet/wallet.cpp]: Using throw without specifying an exception will terminate instead of actually throwing, see below.

Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: yes, you're absolutely right ! I've checked it earlier, that's why I told you that `disable_umask_protection` is necessary there before using `throw`, if I will change `throw` to `fc::exception` subtype then I will remove `disable_umask_protection` call. Just couldn't understand why you told that `disable_umask_protection` isn't necessary before using `throw`.

**Comment by cogutvalera** [comment]
rebased already

**Comment by abitmore** [comment]
Travis-CI complains:
>2393592ms th_a       main.cpp:597                  test_method          ] e.to_detail_string(): 0 exception: unspecified
wallet file cannot be saved
    {}
    th_a  wallet.cpp:824 save_wallet_file 
2393594ms p2p        thread.cpp:252                exec                 ] thread canceled: 9 canceled_exception: Canceled
cancellation reason: [none given]
    {"reason":"[none given]"}
    p2p  thread_d.hpp:473 start_next_fiber
2393594ms th_a       db_management.cpp:206         close                ] Rewinding from 0 to 0
unknown location(0): fatal error in "account_history_pagination": unknown type
/home/travis/build/bitshares/bitshares-core/tests/cli/main.cpp(545): last checkpoint

**Comment by abitmore** [comment]
Perhaps the Travis-CI environment has different `read` behavior or limitations? See https://github.com/bitshares/bitshares-core/pull/1203#issuecomment-408710066 for similar error.

**Comment by cogutvalera** [comment]
going to check ... Thanks !

**Comment by abitmore** [comment]
I'm playing https://github.com/bitshares/bitshares-core/blob/travis-test1/.travis.yml to check if it's related to disk space.

**Comment by cogutvalera** [comment]
wow ! Thanks a lot ! I'm appreciate all your efforts !

**Comment by abitmore** [comment]
Not a disk space issue.
```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.7G  4.0K  3.7G   1% /dev
tmpfs           748M  300K  748M   1% /run
/dev/sda1        30G   18G   11G  62% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
none            5.0M     0  5.0M   0% /run/lock
none            3.7G     0  3.7G   0% /run/shm
none            100M     0  100M   0% /run/user
none            768M  179M  590M  24% /var/ramfs
```

**Comment by cogutvalera** [comment]
rebased already

**Comment by cogutvalera** [comment]
https://travis-ci.org/bitshares/bitshares-core/builds/409736832?utm_source=github_status&utm_medium=notification#L5037

cannot understand why next command `The command "tests/cli_test 2>&1 | cat" exited with 201.` was executed with errors in Travis-CI, locally it works well enough without any errors, does anybody have any  ideas ?

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (pmconrad) [reply libraries/wallet/wallet.cpp]: no, the outer catch should re-throw whatever it caught.

**Comment by cogutvalera** [comment]
ok sure ! Sorry ! Thank you !

File: libraries/wallet/wallet.cpp [hunk 3]
```diff
@@ -808,10 +808,24 @@ class wallet_api_impl
...
          // http://en.wikipedia.org/wiki/Most_vexing_parse
          //
-         fc::ofstream outfile{ fc::path( wallet_filename ) };
+         fc::ofstream outfile{ fc::path( wallet_filename + ".tmp" ) };
          outfile.write( data.c_str(), data.length() );
          outfile.flush();
          outfile.close();
+
+         std::string wallet_file_content;
+         fc::read_file_contents(wallet_filename, wallet_file_content);
```
Comment (abitmore) [inline libraries/wallet/wallet.cpp]: Here you are not reading from the saved file. Why not use a local variable (E.G. `tmp_file_name`)?

Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: you're absolutely right ! made changes already ! Thank you !

**Comment by abitmore** [comment]
Now, when the log shows "Saving wallet to wallet_file_name", actually there are 3 steps:
* save to tmp file
* validate tmp file
* rename tmp to wallet file

However, when the validation fails, the message "can not be saved" doesn't provide useful info.

IMHO it's better to log more useful info, for example, one message per step.

**Comment by cogutvalera** [comment]
```
1955015ms th_a       wallet.cpp:799                save_wallet_file     ] saving wallet to file wallet.json
1955016ms th_a       wallet.cpp:817                save_wallet_file     ] saved successfully wallet to tmp file wallet.json.tmp
1955016ms th_a       wallet.cpp:823                save_wallet_file     ] validated successfully tmp wallet file wallet.json.tmp
1955016ms th_a       wallet.cpp:827                save_wallet_file     ] renamed successfully tmp wallet file to wallet.json.tmp
1955017ms th_a       wallet.cpp:834                save_wallet_file     ] successfully saved wallet to file wallet.json
```

**Comment by abitmore** [comment]
@pmconrad makes sense.

**Comment by cogutvalera** [comment]
#1171 issue concerns about showing any secrets on console, so what is the best approach ? but I also agree with @pmconrad about his idea, it really makes sense and @abitmore was agreed too, let's decide the best approach to be applied everywhere the same way

**Comment by abitmore** [comment]
Wallet file contains no secret. No password, no plain private keys.

**Comment by cogutvalera** [comment]
just for the info wallet file contains next info:

`{
  "chain_id": "4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8",
  "my_accounts": [],
  "cipher_keys": "b30a0ede652bfb2d8d99f8aca77ff8c089f07d77bd047c6df526bc6a652378ed5c37fbbf1d46c408ac1970aec20e83004aee339f1c81e56c0300b00f98f8c142aaad737ddbf7293d339367127a569b1e",
  "extra_keys": [],
  "pending_account_registrations": [],
  "pending_witness_registrations": [],
  "labeled_keys": [],
  "blind_receipts": [],
  "ws_server": "wss://node.bitshares.eu",
  "ws_user": "",
  "ws_password": ""
}`

so really no secret is here, just encrypted keys:

`_wallet.cipher_keys = fc::aes_encrypt( data.checksum, plain_txt );`

**Comment by cogutvalera** [comment]
but what about `"ws_password"` field ? isn't it secret ?

**Comment by cogutvalera** [comment]
in my case `"ws_password"` field is empty so it isn't secret, but it won't always be empty IMHO

**Comment by abitmore** [comment]
I'd say, remove the password before dumping.

File: libraries/wallet/wallet.cpp [hunk 4]
```diff
@@ -838,10 +838,13 @@ class wallet_api_impl
       }
       catch(...)
       {
-         string ws_password = _wallet.ws_password;
-         _wallet.ws_password = "";
-         wlog("wallet file ${data}", ("data", fc::json::to_pretty_string( _wallet ) ) );
-         _wallet.ws_password = ws_password;
+         if (&_wallet != nullptr)
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: &_wallet can never be null. It's a struct member of *this.

Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: it fixed travis error with violation memory access in cli_test as was before, strange behavior of travis that's why I've added this check

Comment (pmconrad) [reply libraries/wallet/wallet.cpp]: The travis error is unrelated, please undo this change.

Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: ok sure

**Comment by abitmore** [comment]
This is still not ready yet. Pushing to next release.

**Comment by cogutvalera** [comment]
can we merge this PR and close related issue ? or this is not still ready and need some changes and improvements ? if YES, then which changes and improvements in your opinion ?

Thanks !

**Comment by pmconrad** [comment]
Thanks!

**Pull request merged**